### PR TITLE
[SPARK-35064][SQL] Group error in spark-catalyst

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
This PR group exception messages in sql/catalyst/src/main/scala/org/apache/spark/sql (except catalyst)

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce any user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

**Comment by beliefer** [comment]
@dgd-contributor Do you check all the `Exception` in package `sql/catalyst/src/main/scala/org/apache/spark/sql` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 1]
```diff
@@ -146,7 +146,7 @@ private[sql] object CatalogV2Util {
                   Some(parent.copy(dataType = addField(parentType, newField, add.position())))
 
                 case _ =>
-                  throw new IllegalArgumentException(s"Not a struct: ${names.init.last}")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:149]: `IllegalArgumentException` is assert-like Exception, please restore it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 2]
```diff
@@ -171,7 +171,7 @@ private[sql] object CatalogV2Util {
         case update: UpdateColumnPosition =>
           def updateFieldPos(struct: StructType, name: String): StructType = {
             val oldField = struct.fields.find(_.name == name).getOrElse {
-              throw new IllegalArgumentException("Field not found: " + name)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:174]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 3]
```diff
@@ -185,7 +185,7 @@ private[sql] object CatalogV2Util {
                 case parentType: StructType =>
                   Some(parent.copy(dataType = updateFieldPos(parentType, names.last)))
                 case _ =>
-                  throw new IllegalArgumentException(s"Not a struct: ${names.init.last}")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:188]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 4]
```diff
@@ -211,7 +211,7 @@ private[sql] object CatalogV2Util {
       val afterCol = position.asInstanceOf[After].column()
       val fieldIndex = schema.fields.indexWhere(_.name == afterCol)
       if (fieldIndex == -1) {
-        throw new IllegalArgumentException("AFTER column not found: " + afterCol)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:214]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 5]
```diff
@@ -224,7 +224,7 @@ private[sql] object CatalogV2Util {
       update: StructField => Option[StructField]): StructType = {
 
     val pos = struct.getFieldIndex(fieldNames.head)
-        .getOrElse(throw new IllegalArgumentException(s"Cannot find field: ${fieldNames.head}"))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:227]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 6]
```diff
@@ -236,15 +236,15 @@ private[sql] object CatalogV2Util {
 
       case (Seq("key"), map @ MapType(keyType, _, _)) =>
         val updated = update(StructField("key", keyType, nullable = false))
-            .getOrElse(throw new IllegalArgumentException(s"Cannot delete map key"))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:239]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 7]
```diff
@@ -236,15 +236,15 @@ private[sql] object CatalogV2Util {
...
+            .getOrElse(throw QueryExecutionErrors.cannotDelete("map key"))
         Some(field.copy(dataType = map.copy(keyType = updated.dataType)))
 
       case (Seq("key", names @ _*), map @ MapType(keyStruct: StructType, _, _)) =>
         Some(field.copy(dataType = map.copy(keyType = replace(keyStruct, names, update))))
 
       case (Seq("value"), map @ MapType(_, mapValueType, isNullable)) =>
         val updated = update(StructField("value", mapValueType, nullable = isNullable))
-            .getOrElse(throw new IllegalArgumentException(s"Cannot delete map value"))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:247]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 8]
```diff
@@ -254,7 +254,7 @@ private[sql] object CatalogV2Util {
 
       case (Seq("element"), array @ ArrayType(elementType, isNullable)) =>
         val updated = update(StructField("element", elementType, nullable = isNullable))
-            .getOrElse(throw new IllegalArgumentException(s"Cannot delete array element"))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:257]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 9]
```diff
@@ -263,8 +263,7 @@ private[sql] object CatalogV2Util {
         Some(field.copy(dataType = array.copy(elementType = replace(elementStruct, names, update))))
 
       case (names, dataType) =>
-        throw new IllegalArgumentException(
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:266]: ditto

**Review by beliefer — COMMENTED** [review]
We no need to group assert-like error like `IllegalArgumentException` or `IllegalStateException`.

**Comment by dgd-contributor** [comment]
> We no need to group assert-like error like `IllegalArgumentException` or `IllegalStateException`.
I found there are many this kind of error in QueryExecutionError.scala, should we restore them too?

**Comment by dgd-contributor** [comment]
> @dgd-contributor Do you check all the `Exception` in package `sql/catalyst/src/main/scala/org/apache/spark/sql` ?

yes, I'm working on checking file by file in the package

**Comment by beliefer** [comment]
> > We no need to group assert-like error like `IllegalArgumentException` or `IllegalStateException`.
> 
> I found there are many this kind of error in QueryExecutionError.scala, should we restore them too?

These are the previous questions, keep them here.

**Comment by dgd-contributor** [comment]
@beliefer can you re-review, I have checked all the exception

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 10]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+      s"Plugin class for catalog '$name' does not implement CatalogPlugin: $pluginClassName")
+  }
+
+  def catalogPluginClassNotFindForCatalogError(name: String, pluginClassName: String): Throwable = {
+    new SparkException(
+      s"Cannot find catalog plugin class for catalog '$name': $pluginClassName")
+  }
+
+  def catalogFailToFindPublicNoArgConstructorError(
+    name: String,
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1443]: Please increase indent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 11]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+    pluginClassName: String,
+    e: Exception): Throwable = {
+    new SparkException(
+      s"Failed to find public no-arg constructor for catalog '$name': $pluginClassName)", e)
+  }
+
+  def catalogFailToCallPublicNoArgConstructorError(
+    name: String,
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1451]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 12]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+    pluginClassName: String,
+    e: Exception): Throwable = {
+    new SparkException(
+      s"Failed to call public no-arg constructor for catalog '$name': $pluginClassName)", e)
+  }
+
+  def catalogCannotInstantiateAbstractPluginClassError(
+    name: String,
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1459]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 13]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+    pluginClassName: String,
+    e: Exception): Throwable = {
+    new SparkException("Cannot instantiate abstract catalog plugin class for " +
+      s"catalog '$name': $pluginClassName", e.getCause)
+  }
+
+  def failInstantiatingConstructorError(
+    name: String,
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1467]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 14]
```diff
@@ -47,35 +48,32 @@ private[sql] object Catalogs {
...
+          .catalogFailToCallPublicNoArgConstructorError(name, pluginClassName, e)
       case e: InstantiationException =>
-        throw new SparkException("Cannot instantiate abstract catalog plugin class for " +
-          s"catalog '$name': $pluginClassName", e.getCause)
+        throw QueryExecutionErrors
+          .catalogCannotInstantiateAbstractPluginClassError(name, pluginClassName, e)
       case e: InvocationTargetException =>
-        throw new SparkException("Failed during instantiating constructor for catalog " +
-          s"'$name': $pluginClassName", e.getCause)
+        throw QueryExecutionErrors
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala:75]: `throw QueryExecutionErrors.failInstantiatingConstructorError(name, pluginClassName, e)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 15]
```diff
@@ -47,35 +48,32 @@ private[sql] object Catalogs {
...
     } catch {
       case _: ClassNotFoundException =>
-        throw new SparkException(
-          s"Cannot find catalog plugin class for catalog '$name': $pluginClassName")
+        throw QueryExecutionErrors.catalogPluginClassNotFindForCatalogError(name, pluginClassName)
       case e: NoSuchMethodException =>
-        throw new SparkException(
-          s"Failed to find public no-arg constructor for catalog '$name': $pluginClassName)", e)
+        throw QueryExecutionErrors
+          .catalogFailToFindPublicNoArgConstructorError(name, pluginClassName, e)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala:67]:
```
QueryExecutionErrors.catalogFailToFindPublicNoArgConstructorError(
  name, pluginClassName, e)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 16]
```diff
@@ -47,35 +48,32 @@ private[sql] object Catalogs {
...
       case e: IllegalAccessException =>
-        throw new SparkException(
-          s"Failed to call public no-arg constructor for catalog '$name': $pluginClassName)", e)
+        throw QueryExecutionErrors
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala:69]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala [hunk 17]
```diff
@@ -47,35 +48,32 @@ private[sql] object Catalogs {
...
+          .catalogFailToCallPublicNoArgConstructorError(name, pluginClassName, e)
       case e: InstantiationException =>
-        throw new SparkException("Cannot instantiate abstract catalog plugin class for " +
-          s"catalog '$name': $pluginClassName", e.getCause)
+        throw QueryExecutionErrors
+          .catalogCannotInstantiateAbstractPluginClassError(name, pluginClassName, e)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/Catalogs.scala:73]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 18]
```diff
@@ -1647,4 +1647,50 @@ private[spark] object QueryCompilationErrors {
...
+
+  def configRemovedInVersion(configName: String, version: String, comment: String): Throwable = {
+    new AnalysisException(
+      s"The SQL config '$configName' was removed in the version $version. $comment")
+  }
+
+  def failedFallbackParsingError(msg: String, e1: Throwable, e2: Throwable): Throwable = {
+    new AnalysisException(
+      message = s"$msg${e1.getMessage}\nFailed fallback parsing: ${e2.getMessage}",
+      cause = Some(e1.getCause))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1659]: `new AnalysisException(s"$msg${e1.getMessage}\nFailed fallback parsing: ${e2.getMessage}", cause = Some(e1.getCause))`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 19]
```diff
@@ -1647,4 +1647,50 @@ private[spark] object QueryCompilationErrors {
...
+  }
+
+  def decimalCannotGreaterThanPrecisionError(scale: Int, precision: Int): Throwable = {
+    new AnalysisException(
+      s"Decimal scale ($scale) cannot be greater than precision ($precision).")
+  }
+
+  def decimalOnlySupportPrecisionUptoError(decimalType: String, precision: Int): Throwable = {
+    new AnalysisException(
+      s"$decimalType can only support precision up to $precision")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1669]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 20]
```diff
@@ -1647,4 +1647,50 @@ private[spark] object QueryCompilationErrors {
...
+      s"You can use spark.sql.legacy.allowNegativeScaleOfDecimal=true " +
+      s"to enable legacy mode to allow it.")
+  }
+
+  def keyNotValidColumnInTableError(key: String, tblName: String): Throwable = {
+    new AnalysisException(s"$key is not a valid partition column in table $tblName.")
+  }
+
+  def partitionSpecNotValidError(
+    specKeys: String,
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1683]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 21]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+  }
+
+  def catalogPluginClassNotImplementError(name: String, pluginClassName: String): Throwable = {
+    new SparkException(
+      s"Plugin class for catalog '$name' does not implement CatalogPlugin: $pluginClassName")
+  }
+
+  def catalogPluginClassNotFindForCatalogError(name: String, pluginClassName: String): Throwable = {
+    new SparkException(
+      s"Cannot find catalog plugin class for catalog '$name': $pluginClassName")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1439]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 22]
```diff
@@ -1422,4 +1423,114 @@ object QueryExecutionErrors {
...
+  }
+
+  def nullLiteralsCannotCastToError(name: String): Throwable = {
+    new UnsupportedOperationException(
+      s"null literals can't be casted to $name")
+  }
+
+  def notUDFError(name: String, userClass: String): Throwable = {
+    new SparkException(
+      s"${name} is not an UserDefinedType. Please make sure registering " +
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1505]: `$name`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 18 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1658]: this line access limit of 100 characters

**Comment by dgd-contributor** [comment]
@beliefer I've resolve all review, is there anything should change, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 23]
```diff
@@ -1647,4 +1647,47 @@ private[spark] object QueryCompilationErrors {
...
+  }
+
+  def partitionSpecNotValidError(
+      specKeys: String,
+      partitionColumnNames: Seq[String],
+      tableName: String): Throwable = {
+    new AnalysisException(
+      s"Partition spec is invalid. The spec ($specKeys) must match " +
+        s"the partition spec (${partitionColumnNames.mkString(", ")}) defined in " +
+        s"table '$tableName'")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1686]:
```
new AnalysisException(s"Partition spec is invalid. The spec ($specKeys) must match " +
  s"the partition spec (${partitionColumnNames.mkString(", ")}) defined in " +
  s"table '$tableName'")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 24]
```diff
@@ -1647,4 +1647,47 @@ private[spark] object QueryCompilationErrors {
...
+  }
+
+  def decimalOnlySupportPrecisionUptoError(decimalType: String, precision: Int): Throwable = {
+    new AnalysisException(s"$decimalType can only support precision up to $precision")
+  }
+
+  def negativeScaleNotAllowedError(scale: Int): Throwable = {
+    new AnalysisException(s"Negative scale is not allowed: $scale. " +
+      s"You can use spark.sql.legacy.allowNegativeScaleOfDecimal=true " +
+      s"to enable legacy mode to allow it.")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1672]: Please refer https://github.com/apache/spark/blob/782dd13d720d0cc3a4799312920c5756ae5e7494/sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala#L986

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 23 above)
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1686]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 25]
```diff
@@ -1422,4 +1423,110 @@ object QueryExecutionErrors {
...
+  }
+
+  def nullLiteralsCannotCastToError(name: String): Throwable = {
+    new UnsupportedOperationException(s"null literals can't be casted to $name")
+  }
+
+  def notUDFError(name: String, userClass: String): Throwable = {
+    new SparkException(
+      s"$name is not an UserDefinedType. Please make sure registering " +
+        s"an UserDefinedType for ${userClass}")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1504]:
```
new SparkException(s"$name is not an UserDefinedType. Please make sure registering " +
  s"an UserDefinedType for $userClass")
```

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1503]: done, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 23 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1685]: done, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 24 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1672]: done, do you mean change spark.sql.legacy.allowNegativeScaleOfDecimal to ${LEGACY_ALLOW_NEGATIVE_SCALE_OF_DECIMAL_ENABLED.key}

**Comment by dgd-contributor** [comment]
@beliefer thanks for the review, I have done what you suggest, can you re-review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 24 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1672]:
```
"""
  | ...
  |...
""".stripMargin.replaceAll("\n", " ")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 23 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1685]:
```
"""
  | ...
  |...
""".stripMargin.replaceAll("\n", " ")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala [hunk 26]
```diff
@@ -45,13 +45,13 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
...
-      s"Decimal scale ($scale) cannot be greater than precision ($precision).")
+    throw QueryCompilationErrors.decimalCannotGreaterThanPrecisionError(scale, precision)
   }
 
   if (precision > DecimalType.MAX_PRECISION) {
-    throw new AnalysisException(
-      s"${DecimalType.simpleString} can only support precision up to ${DecimalType.MAX_PRECISION}")
+    throw QueryCompilationErrors.decimalOnlySupportPrecisionUptoError(
+      DecimalType.simpleString,
+      DecimalType.MAX_PRECISION)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala:54]:
```
throw QueryCompilationErrors.decimalOnlySupportPrecisionUptoError(
  DecimalType.simpleString, DecimalType.MAX_PRECISION)
```

**Review by beliefer — COMMENTED** [review]
cc @allisonwang-db Please check whether the naming reasonable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 27]
```diff
@@ -1647,4 +1647,46 @@ private[spark] object QueryCompilationErrors {
...
+  def decimalCannotGreaterThanPrecisionError(scale: Int, precision: Int): Throwable = {
+    new AnalysisException(s"Decimal scale ($scale) cannot be greater than precision ($precision).")
+  }
+
+  def decimalOnlySupportPrecisionUptoError(decimalType: String, precision: Int): Throwable = {
+    new AnalysisException(s"$decimalType can only support precision up to $precision")
+  }
+
+  def negativeScaleNotAllowedError(scale: Int): Throwable = {
+    new AnalysisException(s"Negative scale is not allowed: $scale. " +
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1670]:
```
"""
  | ...
  |...
""".stripMargin.replaceAll("\n", " ")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 28]
```diff
@@ -1647,4 +1647,46 @@ private[spark] object QueryCompilationErrors {
...
+
+  def keyNotValidColumnInTableError(key: String, tblName: String): Throwable = {
+    new AnalysisException(s"$key is not a valid partition column in table $tblName.")
+  }
+
+  def partitionSpecNotValidError(
+      specKeys: String,
+      partitionColumnNames: Seq[String],
+      tableName: String): Throwable = {
+    new AnalysisException(s"Partition spec is invalid. The spec ($specKeys) must match " +
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1683]:
```
"""
  | ...
  |...
""".stripMargin.replaceAll("\n", " ")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 27 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1670]: why should we use stripMargin.replaceAll("\n", " "), I don't find that in scala-type guide

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1670]: You should follows others.

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: ok :/ Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 28 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: Done, thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 29]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+
+  def noSuchElementExceptionError(key: String): Throwable = {
+    new NoSuchElementException(key)
+  }
+
+  def cannotMutateReadOnlySQLConfError(): Throwable = {
+    new UnsupportedOperationException("Cannot mutate ReadOnlySQLConf.")
+  }
+
+  def cannotCloneCopyReadOnlySQLConfError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1485]: cannotCloneOrCopyReadOnlySQLConfError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 30]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+    new UnsupportedOperationException("Cannot clone/copy ReadOnlySQLConf.")
+  }
+
+  def cannotGetSQLCOnfInSchedulerEventLoopThread(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1489]: cannotGetSQLConfInSchedulerEventLoopThreadError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 31]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+    new RuntimeException("Cannot get SQLConf inside scheduler event loop thread.")
+  }
+
+  def unsupportedOperationExceptionError(): Throwable = {
+    new UnsupportedOperationException
+  }
+
+  def nullLiteralsCannotCastToError(name: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1497]: nullLiteralsCannotBeCastedError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 32]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+    new UnsupportedOperationException(s"null literals can't be casted to $name")
+  }
+
+  def notUDFError(name: String, userClass: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1501]: notUserDefinedTypeError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 33]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+    new SparkException(s"$name is not an UserDefinedType. Please make sure registering " +
+        s"an UserDefinedType for ${userClass}")
+  }
+
+  def cannotLoadUDFError(name: String, userClass: String): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1506]: cannotLoadUserDefinedTypeError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 34]
```diff
@@ -1422,4 +1423,109 @@ object QueryExecutionErrors {
...
+    new SparkException(s"Can not load in UserDefinedType ${name} for user class ${userClass}.")
+  }
+
+  def timestampTypeMustSupplyTZidError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1510]: timeZoneIdNotSpecifiedForTimestampTypeError

**Comment by dgd-contributor** [comment]
@beliefer does sql-slow-test module corrupted? I couldn't know why it keep failing?

**Comment by beliefer** [comment]
retest this please

**Comment by dgd-contributor** [comment]
> @beliefer does sql-slow-test module corrupted? I couldn't know why it keep failing?

my bad, I found out this failing because of extra space in exception message. Thanks.
@allisonwang-db  can you re-review

**Review by allisonwang-db — APPROVED** [review]

**Comment by dgd-contributor** [comment]
@cloud-fan can you check this

**Comment by beliefer** [comment]
@cloud-fan Please help to run tests.

**Comment by dgd-contributor** [comment]
@cloud-fan hi, can you check this

**Comment by dgd-contributor** [comment]
@beliefer since cloud-fan maybe a little bit busy, who else can we request a review? Thanks

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
@dgd-contributor can you update the PR description to follow the template?

**Comment by dgd-contributor** [comment]
> @dgd-contributor can you update the PR description to follow the template?

sure, thanks

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
since github action has passed, I'm merging it to master, thanks!

**Comment by dgd-contributor** [comment]
@beliefer @allisonwang-db thanks for your supporting, I'm appreciated

**Pull request closed without merging**